packages/vite/src/node/utils.ts

```diff
@@ -726,28 +726,29 @@ export interface Hostname {
   name: string
 }
 
+const wildcardHosts = new Set([
+  '0.0.0.0',
+  '::',
+  '0000:0000:0000:0000:0000:0000:0000:0000'
+])
+
 export function resolveHostname(
   optionsHost: string | boolean | undefined
 ): Hostname {
   let host: string | undefined
   if (optionsHost === undefined || optionsHost === false) {
     // Use a secure default
-    host = '127.0.0.1'
+    host = 'localhost'
   } else if (optionsHost === true) {
     // If passed --host in the CLI without arguments
     host = undefined // undefined typically means 0.0.0.0 or :: (listen on all IPs)
   } else {
     host = optionsHost
   }
 
-  // Set host name to localhost when possible, unless the user explicitly asked for '127.0.0.1'
+  // Set host name to localhost when possible
   const name =
-    (optionsHost !== '127.0.0.1' && host === '127.0.0.1') ||
-    host === '0.0.0.0' ||
-    host === '::' ||
-    host === undefined
-      ? 'localhost'
-      : host
+    host === undefined || wildcardHosts.has(host) ? 'localhost' : host
```

Comment:
This is not safe.
When you listen to `0.0.0.0:3000` in Vite, it is still possible to create a server that listens to `127.0.0.1:3000`, and `localhost:3000` will resolve to that server instead.

Reply:
Good catch!
Though I'm not sure how to fix this.
Opening `http://0.0.0.0:3000` in browsers does not work and any other network ip's could also be overridden by other servers.

Maybe adding a warning that denotes ports might be overridden?

Reply:
![image](https://user-images.githubusercontent.com/49056869/173380396-f9f404d3-8ff0-4c84-a0b5-805eced59eb9.png)
I implemented like this. (maybe the message text is not friendly... I didn't come up with a good message.)

Reply:
@sapphi-red I was confused by the message, as you may know. As a Vite user, I would expect it to say "security" and be listed as a warning (not informational). Also while the "ports might be overridden" makes sense in the context of *this* thread, it lead me in a wrong direction.

I know this already shipped, but a suggestion:
"You are using a wildcard host. This means a service listening to `127.0.0.1:port` on your system may hijack your traffic. More information: <link>".

The "more information" should include a way to permanently silence the warning. 

My 2c... Just that you know, anyone running Vite under Docker Compose will need to enable the `host: true` and thus will see the message. Docker Compose port forwarding does not work without it.

*Disclaimer. I don't claim to understand the whole change in play here.*

Reply:
Thanks for the suggestion!

This message is not intended as a security. The purpose of this message is to inform the user that a server other than Vite might response when accessing the output address. Because it is a confusing behavior.

I'll create a PR to improve this message.